### PR TITLE
Add keyword aware quote strategy

### PR DESCRIPTION
**Pull request by Deltachaos**
Related to the discussion in doctrine/dbal#1592 it is probably a good idea to add this quote strategy. It can be enabled by the user optionally and therefore he should be aware of the upercase/lowercase issue in some DBMS.

For the integration of this into doctrine-bundle see this PR: doctrine/DoctrineBundle#553

### CODE REVIEW

**Comment by Ocramius** [comment]
@guilhermeblanco can you take a look at this?

**Comment by Ocramius** [comment]
@Deltachaos tests are missing tho

**Comment by Deltachaos** [comment]
@Ocramius will add tests as soon as is see that there is a change that this gets merged

File: lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     *
+     * @return string
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getQuotedName(AbstractPlatform $platform, $name, $force = false)
+    {
+        $keywords = $platform->getReservedKeywordsList();
+        $parts = explode(".", $name);
+        foreach ($parts as $k => $v) {
+            $parts[$k] = ($force || $keywords->isKeyword($v)) ? $platform->quoteIdentifier($v) : $v;
```
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php]: Unnecessary write of the unmodified value here, rather use normal if.

File: lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+     * @param string $name
+     * @param bool $force
+     *
+     * @return string
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getQuotedName(AbstractPlatform $platform, $name, $force = false)
+    {
+        $keywords = $platform->getReservedKeywordsList();
+        $parts = explode(".", $name);
```
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php]: `"` -> "'"

File: lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+{
+   /**
+     * @param AbstractPlatform $platform
+     * @param string $name
+     * @param bool $force
+     *
+     * @return string
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function getQuotedName(AbstractPlatform $platform, $name, $force = false)
```
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php]: Is there a reason to make this public?

**Comment by Majkl578** [comment]
It'd be really good to have a native way to quote reserved words. 👍

File: lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+
+/**
+ * A set of rules for determining the physical column, alias and table quotes
+ *
+ * @since   2.6
+ * @author  Maximilian Ruta <mr@xtain.net>
+ */
+class KeywordQuoteStrategy implements QuoteStrategy
```
Comment (Majkl578) [inline lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php]: I think _keyword_ is may be a bit misleading, what about e.g. _ReservedWordsQuotingStrategy_?

File: lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php (hunk 3 above)
Comment (Deltachaos) [inline lib/Doctrine/ORM/Mapping/KeywordQuoteStrategy.php]: not realy

**Comment by Deltachaos** [comment]
@Majkl578 tests are added and class was renamed

**Comment by mikeSimonson** [comment]
@Deltachaos Do you still plan on fixing the broken tests?

**Comment by Deltachaos** [comment]
@mikeSimonson yes i have this still on my list. But i can not guarantee to fix this within the next month. Probably start of next year.

**Comment by mikeSimonson** [comment]
@Deltachaos Thanks

**Comment by billschaller** [comment]
Quoting is deeply wtf in DBAL - 3.0 needs to address more than just reserved word quoting...

**Comment by Ocramius** [comment]
@Deltachaos I think we're going for quoting-by-default in ORM 3.x, so I'm not sure if this will be required by then... /cc @guilhermeblanco just to confirm.

**Comment by guilhermeblanco** [comment]
@Ocramius everything is quoted (and cannot be turned off) in `master`.
QuoteStrategy support is also removed from ORM 3.X.

**Comment by Ocramius** [comment]
Closing as `duplicate` then.

**Pull request closed without merging**